proof_ics23.go

```diff
@@ -29,79 +29,52 @@ GetNonMembershipProof will produce a CommitmentProof that the given key doesn't
 If the key exists in the tree, this will return an error.
 */
 func (t *ImmutableTree) GetNonMembershipProof(key []byte) (*ics23.CommitmentProof, error) {
-	// idx is one node right of what we want....
-	var err error
-	idx, val, err := t.GetWithIndex(key)
-	if err != nil {
-		return nil, err
-	}
+	left, err := createExistenceProof(t, key, true)
 
-	if val != nil {
+	if err == nil {
 		return nil, fmt.Errorf("cannot create NonExistanceProof when Key in State")
+	} else if err != ErrKeyNotExist {
+		return nil, err
 	}
 
-	nonexist := &ics23.NonExistenceProof{
-		Key: key,
-	}
+	right, err := createExistenceProof(t, key, false)
 
-	if idx >= 1 {
-		leftkey, _, err := t.GetByIndex(idx - 1)
-		if err != nil {
-			return nil, err
-		}
-
-		nonexist.Left, err = createExistenceProof(t, leftkey)
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	// this will be nil if nothing right of the queried key
-	rightkey, _, err := t.GetByIndex(idx)
-	if err != nil {
+	if err == nil {
+		return nil, fmt.Errorf("cannot create NonExistanceProof when Key in State")
+	} else if err != ErrKeyNotExist {
 		return nil, err
 	}
 
-	if rightkey != nil {
-		nonexist.Right, err = createExistenceProof(t, rightkey)
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	proof := &ics23.CommitmentProof{
+	return &ics23.CommitmentProof{
 		Proof: &ics23.CommitmentProof_Nonexist{
-			Nonexist: nonexist,
+			Nonexist: &ics23.NonExistenceProof{
+				Key:   key,
+				Left:  left,
+				Right: right,
+			},
 		},
-	}
-	return proof, nil
+	}, nil
 }
 
-func createExistenceProof(tree *ImmutableTree, key []byte) (*ics23.ExistenceProof, error) {
-	value, proof, err := tree.GetWithProof(key)
+func createExistenceProof(tree *ImmutableTree, key []byte, leftMost bool) (*ics23.ExistenceProof, error) {
```

Comment:
Just thinking out loud for code hygiene improvement. I think it could make sense to make `createExistenceProof` a func of `ImmutableTree` so you call `t.createExistenceProof(key)`

I think it could make sense to order this function below `GetMembershipProof`, the functions names make sense, since you either want:
- MembershipProof (one existence proof)
- NonMembershipProof(two existence proofs)